coldfront/core/project/views.py

```diff
@@ -1055,93 +1092,147 @@ def dispatch(self, request, *args, **kwargs):
-                            role__name='Principal Investigator').exists():
-                        continue
+class ProjectRemoveSelf(LoginRequiredMixin, UserPassesTestMixin, TemplateView):
 
-                    project_user_obj = project_obj.projectuser_set.get(
-                        user=user_obj)
-                    project_user_obj.status = project_user_removed_status_choice
-                    project_user_obj.save()
+    def test_func(self):
+        if self.request.user.is_superuser:
+            return True
 
-                    # get allocation to remove users from
-                    allocations_to_remove_user_from = project_obj.allocation_set.filter(
-                        status__name__in=['Active', 'New', 'Renewal Requested'])
-                    for allocation in allocations_to_remove_user_from:
-                        for allocation_user_obj in allocation.allocationuser_set.filter(user=user_obj, status__name__in=['Active', ]):
-                            allocation_user_obj.status = allocation_user_removed_status_choice
-                            allocation_user_obj.save()
+        project_obj = get_object_or_404(Project, pk=self.kwargs.get('pk'))
 
-                            allocation_remove_user.send(sender=self.__class__,
-                                                        allocation_user_pk=allocation_user_obj.pk)
+        if project_obj.projectuser_set.filter(
+                user=self.request.user,
+                role__name='User',
+                status__name='Active').exists():
+            return True
 
-            messages.success(
-                request, 'Removed {} users from project.'.format(remove_users_count))
-        else:
-            for error in formset.errors:
-                messages.error(request, error)
+        if project_obj.projectuser_set.filter(
+                user=self.request.user,
+                role__name='Manager',
+                status__name='Active').exists() and \
+                len(project_obj.projectuser_set.filter(role__name='Manager')) > 1:
+            return True
 
-        return HttpResponseRedirect(reverse('project-detail', kwargs={'pk': pk}))
+    def post(self, request, *args, **kwargs):
+        pk = self.kwargs.get('pk')
+        project_obj = get_object_or_404(Project, pk=pk)
+
+        try:
+            request_runner = ProjectRemovalRequestRunner(
+                self.request.user, self.request.user, project_obj)
+            runner_result = request_runner.run()
+            success_messages, error_messages = request_runner.get_messages()
+
+            if runner_result:
+                request_runner.send_emails()
+                for message in success_messages:
+                    messages.success(request, message)
+            else:
+                for message in error_messages:
+                    messages.error(request, message)
+        except Exception as e:
+            self.logger.exception(e)
```

Comment:
This has caused problems for me in the past. This class has no attribute `logger`.

It's safer to define `logger` at the top of the file, and then reference it without this concern.